### PR TITLE
sp-trie: Switch to thiserror and some other small cleanups

### PR DESCRIPTION
**Pull request by bkchr**

### CODE REVIEW

File: primitives/trie/src/storage_proof.rs [hunk 1]
```diff
@@ -121,13 +120,10 @@ impl CompactProof {
 	}
 
 	/// Decode to a full storage_proof.
-	///
-	/// Method use a temporary `HashDB`, and `sp_trie::decode_compact`
-	/// is often better.
```
Comment (bkchr) [inline primitives/trie/src/storage_proof.rs:126]: @cheme I don't really have understood this comment. If you know what this should mean, I can add it back.

Comment (cheme) [reply primitives/trie/src/storage_proof.rs:126]: Yeah, it is not clear. It was a note for user who want to check proof: they should not use this method and check proof from the resulting `StorageProof`.
They should check the proof directly from the `MemoryDb` used by `decode_compact` method.
Generally by looking at the api, one could thing unpack first then check, but that would build the memory db twice.

File: primitives/trie/src/trie_codec.rs [hunk 2]
```diff
@@ -20,80 +20,34 @@
...
+use trie_db::{CError, Trie};
 
 /// Error for trie node decoding.
-pub enum Error<L: TrieConfiguration> {
-	/// Verification failed due to root mismatch.
-	RootMismatch(TrieHash<L>, TrieHash<L>),
-	/// Missing nodes in proof.
+#[derive(Debug)]
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+pub enum Error<H, CodecError> {
```
Comment (cheme) [inline primitives/trie/src/trie_codec.rs:30]: is switching error to `H, CodecError` related to this_error or just design consideration?

File: primitives/trie/src/node_codec.rs [hunk 3]
```diff
@@ -54,9 +54,7 @@ impl<'a> ByteSliceInput<'a> {
 
 impl<'a> Input for ByteSliceInput<'a> {
 	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
-		let remaining =
-			if self.offset <= self.data.len() { Some(self.data.len() - self.offset) } else { None };
-		Ok(remaining)
+		Ok(Some(self.data.len().saturating_sub(self.offset)))
```
Comment (cheme) [inline primitives/trie/src/node_codec.rs:57]: fix code, does not look to me like switching could be an issue :+1:

**Review by cheme — APPROVED** [review]
Nice

File: primitives/trie/src/trie_codec.rs (hunk 2 above)
Comment (bkchr) [reply primitives/trie/src/trie_codec.rs:30]: Related to `Debug` derive :grimacing:

File: primitives/trie/src/node_codec.rs (hunk 3 above)
Comment (bkchr) [reply primitives/trie/src/node_codec.rs:57]: The behavior before was wrong. `remaining_len` should only return `None` if it doesn't know what the remaining length is.

**Pull request merged**